[Proto fuzzer] Add function calls, and multi variable declaration/assignment statements

bshastry opened the pull request:
(closes #6460 )

bshastry committed:
[Proto fuzzer] Add function calls, and multi variable declaration/assignment statements

chriseth commented on test/tools/ossfuzz/yulProto.proto:
```diff
@@ -276,7 +289,13 @@ message Block {
+  required int32 outparams  = 3;
+}
+
+message Program {
```
Shouldn't `Program` be identical to `Block`?

chriseth replied:
Ah, `Statement` does not include `Function` - is that on purpose?

bshastry replied:
`MultiVarAssignment` and `MultiVarDecl`  necessarily include `Function` calls. So, indirectly `Function` call is a statement. `Function` denotes function definition.

chriseth replied:
I'm saying that function definition should be a statement - it can occur almost anywhere in the code.

bshastry replied:
ah, sorry. you're right. Function definition can be a statement. The reason I used the current message hierarchy is to register in advance the list of functions that may later be called.

Changing it may require some code refactoring that I haven't fully thought about. Essentially, we may not know in advance which functions to call until they are visited. I'm also not sure what scoping rules apply to functions. Can they be called before they are defined e.g., 

```
mstore(0, foo())
if expression define foo() -> a { a := 42}
```

chriseth replied:
Scoping rules are the same as we have for variables, just that you can call functions before they are defined. You cannot call them if they are defined in nested scopes, only if they are defined in the current or in outer scopes.

I'm fine with restricting the fuzzer to a certain layout, since the function hoister will pull out all functions rather soon anyway.

bshastry replied:
Postponing making Function definition a statement to a future PR since that might need more code reorganization.

chriseth commented on test/tools/ossfuzz/yulProto.proto:
```diff
@@ -200,6 +205,12 @@ message AssignmentStatement {
   required Expression expr = 2;
 }
 
+message MultiVarAssignmentStatement {
```
Why did you split it into multi var and regular?

bshastry replied:
Mainly because of legacy corpus compatibility and simplicity of conversion logic.

Legacy corpus compatibility means we can directly reuse previous fuzzing corpus which contain regular assignments/decls.

Simplicity of conversion logic means that we can cleanly separate conversion of regular and multi var decl/assignment because a regular statement can only use an expression as RHS but a multi var statement must use a function call as RHS.

chriseth replied:
Ok, that makes sense! From a conceptual standpoint, would it be very bad to add a function call node here instead of directly the function id, params, etc?

chriseth replied:
Oh and function params sholud actually be expressions, shouldn't they?

bshastry replied:
You're right. In general, function parameters are expressions. In this PR, they are variable references that must reference a previously declared expression.

So you are suggesting something like
```
let a := foo(add(x0,x1))
```

at the moment, we have
```
let x2 := add(x0,x1)
let a := foo(x2)
```

I would need to think about how to switch to the former from the current PR :)

bshastry replied:
There is a dependency between `FunctionCall` and `FunctionDefinition` messages that is currently sort of "hard coded" in the conversion logic. I would need to think about how to cleanly separate them so that we no longer hard code them.

bshastry replied:
Added function call message and made function parameters expressions.

bshastry commented:
Update: Incorporated review comments but the code needs to be refactored :)

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,90 @@ class ProtoConverter
+	// Look-up table per function type that holds the number of input (output) function parameters
+	std::vector<int> m_functionVecNoReturnValue;
+	std::vector<int> m_functionVecSingleReturnValue;
+	std::vector<std::pair<int,int>> m_functionVecMultiReturnValue;
```
```suggestion
	std::vector<std::pair<int, int>> m_functionVecMultiReturnValue;
```

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,90 @@ class ProtoConverter
+
+	template<class T>
+	void createFunctionStub(T const&, int, int, NumFunctionReturns);
+	inline std::string functionTypeToString(NumFunctionReturns _type)
```
Why `inline`? (`inline` only has a defined effect on free functions, compilers usually ignore this recommendation)
Also please move to .cpp file.

bshastry replied:
Ah, okay.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,90 @@ class ProtoConverter
+		switch (_type)
+		{
+			case NumFunctionReturns::None:
+				numInputParams = _x.num_input_params() % modInputParams;
```
Can this be moved out of the switch?

bshastry replied:
Indeed, thanks

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
+	registerFunction(_x.fd_zero(), NumFunctionReturns::None);
+	registerFunction(_x.fd_one(), NumFunctionReturns::Single);
+	// A multi return function can have between two and (modOutputParams - 1) parameters
+	int numOutParams = std::max((unsigned int)2, _x.fd_multi().num_output_params() % modOutputParams);
```
```suggestion
	int numOutParams = std::max(2u, _x.fd_multi().num_output_params() % modOutputParams);
```

chriseth replied:
or `std::max<unsigned>(2, ...)`

bshastry replied:
done. To keep things consistent, I changed all "number" types in the ProtoConverter class to unsigned.

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
+	m_output << "{\n";
+	// Create globals at the beginning
+	m_output << "let ";
+	for (int i = 0; i < modOutputParams - 1; i++)
```
Could use `YulUtilFunctions::suffixedVariableNameList`

bshastry replied:
Indeed, it made the code much readable :)

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
+		else
+			m_output << "\n";
+	}
+	if (_x.funcs_size() > 0)
```
Is this required? if `_x.funcs_size() == 0`, doesn't `_x.funcs()` just return an empty list?

bshastry replied:
Yeah, this can be removed.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,90 @@ class ProtoConverter
+	std::vector<int> m_functionVecSingleReturnValue;
+	std::vector<std::pair<int,int>> m_functionVecMultiReturnValue;
+	// mod input/output parameters impose an upper bound on the number of input/output parameters a function may have.
+	static int constexpr modInputParams = 5;
```
Why not `static unsigned constexpr`?

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,90 @@ class ProtoConverter
+	std::vector<int> m_functionVecSingleReturnValue;
+	std::vector<std::pair<int,int>> m_functionVecMultiReturnValue;
+	// mod input/output parameters impose an upper bound on the number of input/output parameters a function may have.
+	static int constexpr modInputParams = 5;
```
Why not `static unsigned constexpr`?

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
-	    << "sstore(32, b)\n"
-	    << "function foo(x_0, x_1, x_2, x_3, x_4, x_5, x_6, x_7) -> x_8, x_9\n";
+	yulAssert(
+			((_numInParams <= modInputParams - 1) && (_numOutParams <= modOutputParams - 1)),
```
Indented too far.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
+}
+
+template <class T>
+void ProtoConverter::createFunctionStub(T const& _x, int _numInParams, int _numOutParams, NumFunctionReturns _type)
```
Does it create the function definition stub or the function call stub?

bshastry replied:
function definition. will rename the function for better readability.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -479,6 +517,128 @@ void ProtoConverter::visit(AssignmentStatement const& _x)
+	// Convert LHS of multi assignment
+	switch (numOutParams)
+	{
+		case 2:
```
These can be combined if you don't use break (order would need to be reversed, but I guess that's fine):
```
case 4:
param4();
case 3:
param3();
...
```

bshastry replied:
Ah, very cool :)

ekpyron replied:
I hope our compiler settings won't prevent that (the "missing" break might trigger a warning and our settings might consider that warning an error.). With C++17 fallthroughs could be explicitly marked :-): https://en.cppreference.com/w/cpp/language/attributes/fallthrough - we should probably move forward with the move to C++17.

bshastry replied:
done. Had to annotate with `BOOST_FALLTHROUGH;` otherwise clang complains of implicit fallthrough.

bshastry commented:
Updated the PR. Changes from last review are

- used `YulUtilFunctions::suffixedVariableNameList` instead of custom for loop (introduces libsolidity as dependency, see #6710 )
  - Added libsolidity as dependent library for the non-differential proto fuzzer since the differential proto fuzzer already transitively links libsolidity via libyulInterpreter
- made switch cases more elegant where possible (used BOOST_FALLTHROUGH annotation to silence compiler warning)
- Increased test file size limit from 600 bytes to 1200 bytes
  - This is experimental and an optimal upper bound needs to be experimentally determined
  - Increase was necessary because the fuzzer was more regularly synthesizing larger test cases. This is expected because we have at least 3 functions in each test.

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -226,6 +232,43 @@ void ProtoConverter::visit(VarDecl const& _x)
+	// We must start variable numbering past the number of live variables at this point in time.
+	// This creates let x_p,..., x_k :=
+	// (k-p)+1 = numOutParams
+	m_output << "let "
```
`<<` should always be the last token on the line and the lines below should be indented one tab less than they currently are.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -479,6 +522,115 @@ void ProtoConverter::visit(AssignmentStatement const& _x)
+	// We reverse the order of function input visits since it helps keep this switch case concise.
+	switch (_numInputParams)
+	{
+		case 4:
```
We usually do not indent the case labels.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -479,6 +522,115 @@ void ProtoConverter::visit(AssignmentStatement const& _x)
+	// switch statement.
+	switch (numOutParams)
+	{
+		case 4:
```
Same here.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -479,6 +522,115 @@ void ProtoConverter::visit(AssignmentStatement const& _x)
+{
+	switch (_x.functioncall_oneof_case())
+	{
+		case FunctionCall::kCallZero:
```
And here.

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +834,173 @@ void ProtoConverter::visit(Block const& _x)
+	m_output << "{\n";
+	visit(_x.var());
+	if (_x.statements_size() > 0)
+	{
```
No need for the `{`

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +834,173 @@ void ProtoConverter::visit(Block const& _x)
+	// Manually create a multi assignment using global variables
+	// This prints a_0, ..., a_k-1 for this function that returns "k" values
+	if (_numOutParams > 0)
+	{
```
No need for the `{`

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +834,173 @@ void ProtoConverter::visit(Block const& _x)
+	m_output << ")\n";
+
+	for (unsigned i = 0; i < _numOutParams; i++)
+	{
```
No need for the `{`

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +834,173 @@ void ProtoConverter::visit(Block const& _x)
+	std::string ret;
+	switch (_type)
+	{
+		case NumFunctionReturns::None:
```
Please unindent.

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.cpp:
```diff
@@ -677,28 +834,173 @@ void ProtoConverter::visit(Block const& _x)
+	switch (_type)
+	{
+		case NumFunctionReturns::None:
+			ret = "noreturn";
```
Why not just `return "noreturn";`?

bshastry replied:
done

chriseth commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -75,19 +80,72 @@ class ProtoConverter
+		unsigned numInputParams = _x.num_input_params() % modInputParams;
+		switch (_type)
+		{
+			case NumFunctionReturns::None:
```
Please unindent.

bshastry replied:
done.

bshastry commented:
Thank you. I unindented all case statements :)
